Propagate labels across all repos

lukaszgryglicki opened the pull request:
This PR adds `label_sync` tool: #2504 
It uses `labels.yaml` file to sync standard labels set across all `kubernetes` repos.
Please refer to `label_sync/README.md` for details.

This PR is *NOT* ready - I've created new tool, but it is not connected. I needs to be run `daily` like something like `prow` job, but I didn't yet have a time to figure out how to connect it. It is also my first task like it so *ANY* feedback about how to connect it is very very very welcomed.

Main souce files are: `label_test/main.go` and `label_sync/main_test.go`.

krzyzacy commented:
@k8s-bot ok to test

/assign @spxtr @grodrigues3

grodrigues3 commented:
Let's incorporate Brian's feedback.  Get rid of the `team/` labels.  And add:

```
cncf-cla*
approved
do-not-merge
needs-ok-to-merge
needs-ok-to-test
needs-rebase
ok-to-merge
queue/*
retest-*
size/*
```

spxtr commented on label_sync/repos.yaml:
```diff
@@ -0,0 +1,242 @@
+Org: kubernetes
```
Why do we need this file? Ideally the labels syncer can just do a github list repos call to sync all repos under the org. That way we don't need to update this file when we add a new repo.

spxtr replied:
Oh I see, it's documented elsewhere. Do we want to check in this file?

lukaszgryglicki replied:
It is used to help local development mode. You can save github points using this file instead of doing real Github API call.

spxtr commented on label_sync/repo_labels.yaml:
```diff
@@ -0,0 +1,3527 @@
+  - color: ffaa04
+    name: area/extensibility
+    url: https://api.github.com/repos/kubernetes/kube-ui/labels/area/extensibility
+  -
```
:)

This one is very important.

lukaszgryglicki commented:
@grodrigues3 addressed Your feedback: https://github.com/kubernetes/test-infra/pull/2706#issuecomment-300569667

grodrigues3 commented:
@lukaszgryglicki have you joined the kubernetes-dev mailing list  There is a [discussion](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubernetes-dev/0rTJ6wjcitg/zDCv-kPBAAAJ) about this work.  Can you review it and respond with regards to

-  which labels we will use (suggestions we only use the sig ones and a few others to start)
-  which repos they will be applied to (people seem to not want labels propagated across all repos)

lukaszgryglicki commented:
Well if we don't want to propagate across all repos, then we can call program with `-repos file_name.yaml` to restrict list of repositories.

lukaszgryglicki commented on label_sync/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+WORKDIR /go/src/k8s.io/test-infra
+ADD ./label_sync/ /go/src/k8s.io/test-infra/label_sync/
+ADD ./labels.yaml /etc/config/labels.yaml
+# ADD ?secret? /etc/github/oauth
```
How to add our repo's secret Github OAuth2 token?
Also please not that this Dockerfile is VERY hip. I need help with that part!
@grodrigues3 @spxtr ?

kargakis replied:
Should be passed as a secret. See how this is done in prow.

https://github.com/kubernetes/test-infra/blob/0ad913aef9819fda96252883df0987ef5382cabe/prow/cluster/hook_deployment.yaml#L63

lukaszgryglicki replied:
I will try to follow it once I finish syncing `labels.yaml` file.
BTW: what is the general strategy to create such programs?
I can:
1. Create smallest possible docker image with minimal Linux and only `label_sync` binary.
2. Create YAML definition that uses this image and only connects secrets needed, just like the YAML You linked.

Or maybe I should not compile binary, but create a bigger image with golang installed there, and it will compile program within image?

lukaszgryglicki commented on label_sync/README.md:
```diff
@@ -0,0 +1,40 @@
+`go run label_sync/main.go`
+
+
+TODO:/FIXME:
```
FIXME/TODO w.i.p

spxtr commented:
I'm gone for the next two weeks. I think @grodrigues3 can help review this code. If you have further questions about setting up prow jobs and credentials and such, please ask @rmmh.

krzyzacy commented on label_sync/BUILD:
```diff
@@ -0,0 +1,59 @@
+package(default_visibility = ["//visibility:public"])
```
run `./verify/update-bazel.sh` to fix the verify-bazel test.

lukaszgryglicki replied:
Done.

krzyzacy commented:
@lukaszgryglicki https://github.com/kubernetes/test-infra/blob/master/prow/cluster/hook_deployment.yaml#L50 is a good reference, we already bind github oauth token to prow.

rmmh commented on label_sync/repo_labels.yaml:
```diff
@@ -0,0 +1,3527 @@
+Labels:
+  apimachinery:
```
is this file purely machine generated? If we want humans to edit it, it might be friendlier to have a format like:

    apimachinery:
    - "bug #ee0701"
    - "cncf-cla: no #ededed"

etc

rmmh replied:
And I don't think the URLs are interesting at all

lukaszgryglicki replied:
This file is machine generated and only needed for local testing without consuming GitHub API points.

lukaszgryglicki replied:
Removed because it was for testing only.

rmmh commented on label_sync/README.md:
```diff
@@ -0,0 +1,40 @@
+You can run it locally with repositories and lables data fetched into YAML files:
```
s/lables/labels/

rmmh commented on labels.yaml:
```diff
@@ -0,0 +1,284 @@
+  - color: fbca04
+    name: kind/upgrade-test-failure
+  - color: 0052cc
+    name: area/admin
```
area/ labels probably aren't globally interesting-- lots of these components/modules don't exist in any meaningful way outside of kubernetes/kubernetes

lukaszgryglicki replied:
I don't know which ones are interesting and which ones aren't. I've just added lables requested.
Actually labels.yaml will probably be a lot shorter for a start. And `repos.yaml` file will be used instead of getting all kubernetes repos via GitHub call (because we want to sync only few repos for a start). so those two files: `labels.yaml` and `repos.yaml` will have to be tweaked to some wide accepted consensus.

lukaszgryglicki replied:
I've generated new `labels.yaml` file from most recent `kubernetes/kubernetes` and I'm addressing feedback by commenting out labels that are not needed IMHO. I'm not removing them, but comment out to confirm I'm OK or wrong with particular ones.

rmmh commented on labels.yaml:
```diff
@@ -0,0 +1,284 @@
+  - color: fbca04
+    name: ok-to-merge
+  - color: ffaa00
+    name: queue/blocks-others
```
what are these? does anything actually consume these?

lukaszgryglicki replied:
Those are labels `queue/*` from taken from `kubernetes/kubernetes` repo. I was requested to sync them too.

bgrant0607 replied:
The queue labels were created to replace the numerical priority labels for submit queue ordering.

lukaszgryglicki replied:
So I don't know if I should remove them or not, but at least I've synced `labels.yaml` with current kubernetes/kubernetes.

lukaszgryglicki commented:
@krzyzacy - I was requested to rather use new oath token for this - other than token used by `prow` - how can I generate it? Or who can do it?
Thanks.
Also can You point me to some "standards" about generating Dockerfiles for such new programs?
Should I create image with `godev` and compile `label_sync` there?
Or should I compile it myself on some ubuntu image and finally create Dockerfile with just plain ubuntu + compiled label_sync + new oath token and config files?

lukaszgryglicki commented:
I don't know why those bazel tests fail.
Locally they're OK if I modify file: `vendor/github.com/Sirupsen/logrus/terminal_notwindows.go`
Via daaing those lines:
```
type Termios syscall.Termios
const ioctlReadTermios = syscall.TIOCGETA
```

But this is only needed on Mac imho.
More fun: running: `go test k8s.io/test-infra/label_sync` to test the same without bazel doesn't need any changes to `terminal_notwindows.go` file

I've did `verify/update-bazel.sh`.
Actual errors linked above are unrelated imho:
```
  Clusters refilter filters by test:
     TypeError: Cannot read property 'length' of undefined
      at Clusters.refilter (/workspace/k8s.io/test-infra/triage/model.js:182:19)
      at Context.<anonymous> (/workspace/k8s.io/test-infra/triage/script_test.js:33:36)
```

krzyzacy commented:
@k8s-bot bazel test this

krzyzacy commented:
The image should just include the program, a sample prow Dockerfile is like https://github.com/kubernetes/test-infra/blob/master/prow/cmd/horologium/Dockerfile, and built from https://github.com/kubernetes/test-infra/blob/master/prow/Makefile#L49-L52.

The auth token should be a prow secret. It's for k8s-ci-bot I think. It has 2FA on, I can try but maybe I'll need one of @spxtr or @ixdy to be back.. Let us know what scope of access token you'll need.

grodrigues3 commented:
@krzyzacy I'll let you LGTM whenever you think it's ready

krzyzacy commented:
/unassign
/assign @spxtr 

this PR is back to you sir :-)

spxtr commented:
OK I'll take a look.

grodrigues3 commented:
ping @lukaszgryglicki @spxtr 

When can we get this merged?

bgrant0607 commented on labels.yaml:
```diff
@@ -0,0 +1,284 @@
+  - color: d2b48c
+    name: sig/apps
+  - color: d2b48c
+    name: sig/auth
```
This list isn't up to date. It doesn't contain sig/architecture, for instance.

lukaszgryglicki replied:
Synced with current `kubernetes/kubernetes` so now it does.

bgrant0607 commented on labels.yaml:
```diff
@@ -0,0 +1,284 @@
+  - color: eb6420
+    name: priority/important-soon
+  - color: ff0000
+    name: priority/P0
```
We shouldn't copy these deprecated labels to other repos. Similar for team labels.

lukaszgryglicki replied:
Commented out.

spxtr commented:
Because this needs access to an oauth token that has write access to the kubernetes org, it can't run as a ProwJob. Instead, it will have to run continuously in either the prow or mungegithub cluster.

lukaszgryglicki commented:
Sorry guys I wasn't working on this since May. And today I have a lot of other urgent tasks to do.
I think it was almost finished, but was missing images and proper "connection" to the rest of the system.
This was just because it was my first task with Kubernetes.
I will get back to it when I can, but actually, I'm not sure how to proceed with it.
Labels list can, of course, be outdated because it was generated at the time of PR creation.

spxtr commented:
I think @grodrigues3 wants this to go in ASAP, so would you mind if someone else picks it up from you? We can use your work to start off.

grodrigues3 commented:
Yeah, I'd really like to see this get in this week.  Let's clearly define the work that needs to be done, and I'll look for a way to get it staffed.

lukaszgryglicki commented:
@spxtr Yes, of course, someone else can continue this. To be honest my load average last week and this one is just about 3.0.

kargakis commented on label_sync/main.go:
```diff
@@ -0,0 +1,438 @@
+	labelsPath         = flag.String("labels-path", "/etc/config/labels.yaml", "Path to labels.yaml.")
+	githubTokenFile    = flag.String("github-token-file", "/etc/github/oauth", "Path to the file containing the GitHub OAuth secret.")
+	org                = flag.String("org", "kubernetes", "Organization to fetch repository list from, to fetch user's public repositories, use user:user_name")
+	dry                = flag.Bool("dry", false, "Dry run for testing. Uses API tokens but does not mutate.")
```
dry-run for consistency with elsewhere

lukaszgryglicki replied:
Done.

kargakis commented on label_sync/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+RUN git clone https://github.com/kubernetes/test-infra
+WORKDIR /go/src/k8s.io/test-infra
+ADD ./label_sync/ /go/src/k8s.io/test-infra/label_sync/
+ADD ./labels.yaml /etc/config/labels.yaml
```
This file should be passed as a configmap inside the pod otherwise we will need to rebuild with every change.

lukaszgryglicki replied:
Not sure how to do that, but I will have a discussion on Hangouts today 10:30 am PDT. I'm syncing `labels.yaml` and then I'll try to create minimal docker image/Dockerfile. Then I will create YAML with that image connected and also with secrets connected.
But this is my first time writing some Kubernetes code, so I have no idea how to do it actually.

lukaszgryglicki commented:
I will at least try to rebase it tomorrow. I don't have more time for working on this atm.

lukaszgryglicki commented:
I've also tried to address deployment issues - but I have no idea how should it work finally.
Please - anybody who reviews it - take a look at label_sync/:
- `Dockerfile`
- `label_sync_deployment.yaml`

There are ton of questions there - marked with `TODO:`.

lukaszgryglicki commented:
I've done all I could to proceed on this, and now all depends on answers to my questions:
Here are my outstanding questions (I can work on this tomorrow all day if I receive feedback allowing me to do so, or someone else can pick this up I think it is VERY close to finish - it only needs deployment part, docker image, pod definition, cron job definition).

I don't know how to deploy apps in kubernetes/test-infra.

1) Docker image (docker file)
-Should I compile `label_sync` into go binary and just put compiled binary in the image? Or should I copy sources into the container and compile them inside the container?
-When is Docker image generated? I want to try full workflow locally?
-How should I name the image and where it is configured? (for now I can `docker build` and `docker run` manually)
-Please take a look at `Dockerfile`'s TODO's

2) Pod definition (like for hook deployment suggested on the PR):
-When the pod is generated, what is the full workflow?
-Should it have single replica right? (because it is supposed to run periodically and update all Kubernetes repos labels)
-It lists image to use here - the name is in format `k8s.io...label_sync` - where should I define this image?
-It is a standalone program, not the `prow` plugin, and it should run periodically say once per day - where can I define such a cron job?
-Is my `oauth secret` correctly defined there?
-Please take a look at `label_sync/label_sync_deployment.yaml`'s TODO's

3) `label_sync/labels.yaml` file:
-It was outdated because it was generated 15th May 2017. I've fetched all labels from  kubernetes/kubernetes today and generated new file - then I've reviewed all PR feedback and commented out labels that are not needed - it just needs review. Maybe I should remove a lot more labels?

spxtr commented on label_sync/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# TODO: alpine is smaller, but if I compile Go binary locally it won't run there
```
Compile with `CGO_ENABLED=0`, then it should work.

kargakis replied:
Don't you also need to statically compile?

lukaszgryglicki replied:
It works by running the command: `CGO_ENABLED=0 go build ./main.go`
`ldd` says compiled binary is static, and now works with `alpine` image. Good.

spxtr commented on label_sync/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+FROM golang
+MAINTAINER lukaszgryglicki@o2.pl
+
+# TODO: What this is for - images are supposed to be minimal, I think it is not needed
```
This is needed if you're talking to someone over HTTPS, such as `https://api.github.com/` :)

lukaszgryglicki replied:
That was my thought too, so removing this line. Thanks.

spxtr commented on label_sync/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+
+# TODO: this is supposed to be done as secret configured in YAML
+# RUN mkdir /etc/github
+# ADD oauth /etc/github/oauth
```
Nope, these will be added by Kubernetes.

lukaszgryglicki replied:
OK. I've added this to my YAML config, but have no idea how to test it locally - that's my biggest concern atm.

spxtr commented on label_sync/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+
+# TODO: this copies compiled binary into image
+ADD main /label_sync
+ADD labels.yaml /etc/labels.yaml
```
Instead, we can add these with kubernetes by making them a configmap.

spiffxp replied:
This would be fantastic, it's helped reduce roundtrip time for other changes

lukaszgryglicki replied:
How to do this, and where to configure this?

lukaszgryglicki replied:
Hey - this `Makefile` was EXACTLY what I was looking for - now I can see the workflow. THANKS!!

lukaszgryglicki replied:
For now, I will leave adding `labels.yaml` "as is" - it is just a single file. If we really want to add it somewhat different than I need instructions how to do so.
GitHub OAuth secret should already be handled in `label_sync_deployment.yaml`.

spxtr commented:
> Should I compile `label_sync` into go binary and just put compiled binary in the image? Or should I copy sources into the container and compile them inside the container?

I've been doing the former for prow components. See, for instance, hook's [Dockerfile](https://github.com/kubernetes/test-infra/blob/103d9a331b9c7e9bbe882da6b4eaa4e2e5aabffd/prow/cmd/hook/Dockerfile) and [Makefile entry](https://github.com/kubernetes/test-infra/blob/103d9a331b9c7e9bbe882da6b4eaa4e2e5aabffd/prow/Makefile#L58-L61).

> When is Docker image generated? I want to try full workflow locally?
> How should I name the image and where it is configured? (for now, I can docker build and docker run manually)

See that makefile.

> Please take a look at Dockerfile's TODO's

Done. For my comments that say "we can do with with a kubernetes configmap", an example deployment is [here](https://github.com/kubernetes/test-infra/blob/103d9a331b9c7e9bbe882da6b4eaa4e2e5aabffd/prow/cluster/hook_deployment.yaml#L44-L69) that mounts a configmap. For local testing, mount them with a docker volume.

spxtr commented on label_sync/label_sync_deployment.yaml:
```diff
@@ -0,0 +1,52 @@
+  labels:
+    app: label_sync
+spec:
+  replicas: 1 # TODO I assume 1 because it must run alone
```
Correct.

spxtr commented on label_sync/label_sync_deployment.yaml:
```diff
@@ -0,0 +1,52 @@
+        - -github-bot-name=k8s-ci-robot
+        - -dry-run
+        - -debug
+        ports:   # TODO This is not needed at all for this pod
```
Correct, feel free to remove it.

spxtr commented on label_sync/label_sync_deployment.yaml:
```diff
@@ -0,0 +1,52 @@
+      terminationGracePeriodSeconds: 180
+      containers:
+      - name: label_sync
+        image: gcr.io/k8s-label-sync   # TODO Where this image is supposed to be generated, how should I name it
```
We can put it in `gcr.io/k8s-prow`.

spxtr commented:
> When the pod is generated, what is the full workflow?

When you make the cronjob, k8s will make the pod for you.

> Should it have single replica right? (because it is supposed to run periodically and update all Kubernetes repos labels)

Correct.

> It is a standalone program, not the prow plugin, and it should run periodically say once per day - where can I define such a cron job?

A kubernetes deployment will be running all the time. If you want it to run once daily, use a kubernetes [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/).

> Sorry if some questions seems stupid - but this is my first contribution to kubernetes, especially in deployment area that is a total black magic for me atm.

No worries at all. Kubernetes has a learning curve and we asked to to take on a challenging task right away. What you have already is pretty close to complete :)

spiffxp commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+  - color: e11d21
+    name: do-not-merge
+  - color: fbca04
+    name: flake-has-meta
```
flake-has-meta can be commented out

(this should probably be outright removed)

spiffxp commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+  - color: fbca04
+    name: flake-has-meta
+  - color: 006b75
+    name: for-new-contributors
```
for-new-contributors can be commented out

(I would eventually like to see this standardized, but my aim for this pass is standard automation labels)

spiffxp commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+  - color: 006b75
+    name: for-new-contributors
+  - color: 006b75
+    name: help-wanted
```
help-wanted can be commented out

(I would eventually like to see this standardized, but my aim for this pass is standard automation labels)

spiffxp commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+  - color: 15dd18
+    name: lgtm
+  - color: ededed
+    name: needs-ok-to-merge
```
needs-ok-to-merge can be commented out

(this should be outright removed, this mungegithub functionality is gone)

spiffxp commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+  - color: ededed
+    name: needs-sig
+  - color: 0e8a16
+    name: non-release-blocker
```
non-release-blocker can be commented out

(this should probably be removed)

spiffxp commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+#  - color: eb6420
+#    name: priority/important-longterm
+#  - color: eb6420
+#    name: priority/important-soon
```
I would uncomment all of these non-numeric `priority/*` labels, they're [well documented](https://github.com/kubernetes/community/blob/master/contributors/devel/issues.md#define-priority)

spiffxp commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+  - color: ffaa00
+    name: queue/multiple-rebases
+  - color: d93f0b
+    name: release-blocker
```
I would comment out release-blocker

(we should probably get rid of this label, presence in the release milestone == blocking the release)

spiffxp commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+  - color: eb6420
+    name: retest-not-required
+  - color: fbca04
+    name: retest-not-required-docs-only
```
more of a note-to-self: these `retest-not-required` labels are sorted/compared alongside the `queue/*` labels, maybe they should be prefixed with `queue/` as well?

spiffxp commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+  - color: ee0000
+    name: size/XXL
+  - color: "795548"
+    name: stale
```
stale can be commented out

(this should probably be removed, or we should find criteria to have the bot auto-apply this and tack a prefix on it)

lukaszgryglicki replied:
Addressed all labels related feedback. I didn't removed any labels - I'm only commenting or uncommenting - I want to have full labels set for further/final discussion - so any label that should be removed is kept commented out atm.

spiffxp commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+#  - color: c7def8
+#    name: kind/technical-debt
+#  - color: fbca04
+#    name: kind/upgrade-test-failure
```
OK, bikeshed time!  The full list of `kind/*` labels is gratuitous, but it would be nice if the `/kind` command was useful across all kubernetes repos. 

I propose that the following subset get uncommented:
- `kind/bug`
- `kind/documentation`
- `kind/feature`

lukaszgryglicki replied:
We can consider starting with as small as possible.
Why?
This tool is only syncing labels listed there:
Adds missing or changes the color of existing.
It is not removing any labels, so labels not listed there will not be touched on other repos at all.

alexandercampbell replied:
+1 to starting as small as possible and doing the `kind/{bug,doc,feature}` in a different PR.

spiffxp commented on label_sync/main_test.go:
```diff
@@ -0,0 +1,338 @@
+	"strings"
+	"testing"
+
+	"k8s.io/test-infra/prow/github"
```
as we make changes to prow's codebase, will a presubmit job catch if it breaks assumptions here?

lukaszgryglicki replied:
Yes, but using prow's one I can use FakeClient, DryRunClient and normal client - so this is very useful.

spiffxp commented on label_sync/main.go:
```diff
@@ -0,0 +1,438 @@
+	"github.com/Sirupsen/logrus"
+	"github.com/ghodss/yaml"
+
+	"k8s.io/test-infra/prow/github"
```
as we make changes to prow's codebase, will a presubmit job catch if it breaks assumptions here?

lukaszgryglicki replied:
No idea.

spxtr replied:
Yes.

spiffxp commented on prow/github/client.go:
```diff
@@ -435,6 +435,70 @@ func (c *Client) CreateStatus(org, repo, ref string, s Status) error {
 }
 
+func (c *Client) GetRepos(org string, is_user bool) ([]Repo, error) {
+	c.log("GetRepos", org)
```
also log is_user

spiffxp commented on label_sync/single_repo_example.yaml:
```diff
@@ -0,0 +1,8 @@
+Org: kubernetes
+Repos:
```
can these be `org` and `repos` instead?

lukaszgryglicki replied:
That file is not needed - it is just a helper to get newest Kubernetes labels file (for example when we need a fresh copy of labels given day to update our `labels.yaml` file).
This file is generated by Marshalling data into YAML (and fields are an upper case because they need to be public):
```
// RepoList is a list of Repositories from Org to be checked by this program
type RepoList struct {
        Org   string
        Repos []github.Repo
}
```
If they were a lower case they won't be accessible in go code then.
So without messing with this and adding complexity to the code, I will leave it "as is".
Other option can be to remove that file totally. It is not used for normal workflow. It is only a local mode helper if You want to get newest label's set from a given org/repo list.

spxtr replied:
```go
// RepoList is a list of Repositories from Org to be checked by this program
type RepoList struct {
        Org   string        `json:"org"`
        Repos []github.Repo `json:"repos"`
}
```
:)

spxtr replied:
I would rather remove the file from the PR and add it to `.gitignore`, people can generate it locally if they'd like.

lukaszgryglicki replied:
Will do. It is not needed at all.

lukaszgryglicki commented:
@spxtr :
```
A kubernetes deployment will be running all the time. If you want it to run once daily, use a kubernetes CronJob.
```

I'm wondering what does it mean `will be running all the time`.
If that means it will just run label_sync each time Kubernetes is deployed (I don't know how often this happen) - maybe we're just OK with it?
Of course, I don't want it to run all the time meaning when one instance finishes it runs another - it is too often and will waste GitHub API points.

I'll read Kubernetes `CronJob` docs now -  and depending on what I find there - I'll eventually create CronJob.
I'll push all my changes after that - in a few hours max.

All other feedback is hopefully addressed, this missing part was that `Makefile` that enlighted me how the things works.

lukaszgryglicki commented:
Addressed feedback and added Cron job - needs to be reviewed.

lukaszgryglicki commented on label_sync/main_test.go:
```diff
@@ -0,0 +1,338 @@
+	"strings"
+	"testing"
+
+	"k8s.io/test-infra/prow/github"
```
No idea.

spxtr replied:
Yes, but I'd prefer you use `go-github`. Prow's github client was intended to be used by prow alone...

lukaszgryglicki commented:
@spxtr what do You think?

spxtr commented:
> I'm wondering what does it mean will be running all the time.
> If that means it will just run label_sync each time Kubernetes is deployed (I don't know how often this happen) - maybe we're just OK with it?
> Of course, I don't want it to run all the time meaning when one instance finishes it runs another - it is too often and will waste GitHub API points.

It will make sure that your program is running constantly. If it finishes and closes then kubernetes will bring it back up. This is probably not what you want :)

lukaszgryglicki commented:
Yes, that's true.
I've also created CronJob (at least I was trying).
So now I will have to make sure it is *NOT* run by anything else than cron.

lukaszgryglicki commented:
I don't know why `gofmt` failed. I've did changes to main.go (to add `json "org"` etc.) but I've also run gofmt -w on label_sync/main.go.
So this should be OK.
/retest

spxtr commented:
We started using `gofmt -s` which does a few more simplifications.

lukaszgryglicki commented:
OK, I'll check sources now and push again

spxtr commented:
/lint

spxtr commented on label_sync/main.go:
```diff
@@ -0,0 +1,438 @@
+limitations under the License.
+*/
+
+package main
```
Feel free to ignore this one.

lukaszgryglicki commented:
Ohh lint, one more round then.

lukaszgryglicki commented:
All tests green.

grodrigues3 commented:
@spxtr

grodrigues3 commented:
@lukaszgryglicki @spiffxp @spxtr 

Looks like we're pretty close to ready to merge. 

@spiffxp are we good with the subset of labels to be used? 

Before we run this, we should send an email to kubernetes-dev@.

spiffxp commented on label_sync/main.go:
```diff
@@ -0,0 +1,439 @@
+	labelsPath         = flag.String("labels-path", "/etc/config/labels.yaml", "Path to labels.yaml.")
+	githubTokenFile    = flag.String("github-token-file", "/etc/github/oauth", "Path to the file containing the GitHub OAuth secret.")
+	org                = flag.String("org", "kubernetes", "Organization to fetch repository list from, to fetch user's public repositories, use user:user_name")
+	dry                = flag.Bool("dry-run", false, "Dry run for testing. Uses API tokens but does not mutate.")
```
sorry for noticing this so late... we might want to default this to true, to follow a pattern set by mungegithub and prow

spiffxp commented:
@grodrigues3 the labels LGTM, and I would be fine merging as is

But now that I'm looking at this from the perspective of rolling it out, I'm wary of just flipping a switch and saying go

Are we planning on selectively turning this on for a repo or two before hitting the whole org?

grodrigues3 commented:
@spxtr i'll let you give the final lgtm.

@spiffxp running on one or two repos seems the prudent thing to do.  test-infra (we can monitor) and contrib (not used so much right now)?

lukaszgryglicki commented:
Changed dry-run to be true by default.
Please note that in this case default deployment will run in dry-mode, effectively doing nothing. But this way we can see if all deployments and cron job works. If so, then we can change deployments to include flag `--dry-run=false`.
To limit running to single repo, like `kubernetes/test-infra` we need to use `-repos file.yaml` flag, and `file'yaml` must contain list of repos to run, in this case only one repo: `kubernetes/test-infra`, like this:
```
org: kubernetes
repos:
- full_name: kubernetes/test-infra
  name: test-infra
```

lukaszgryglicki commented:
Does anything happen on this?
I think it is approved, what's next?

alexandercampbell commented on label_sync/Makefile:
```diff
@@ -0,0 +1,59 @@
+all: build test
+
+
+LABEL_SYNC_VERSION       ?= 0.100
```
What's up with this version number? I've never seen this format before.

lukaszgryglicki replied:
I'll change to 0.1

alexandercampbell commented on label_sync/Makefile:
```diff
@@ -0,0 +1,59 @@
+REGISTRY ?= gcr.io
+PUSH     ?= gcloud docker -- push
+
+DOCKER_LABELS=--label io.k8s.prow.git-describe="$(shell git describe --tags --always --dirty)"
```
I'd prefer to see this without the `--label` as part of the variable and later do

```shell
cmd --labels "$DOCKER_LABELS"
```

lukaszgryglicki replied:
I was following example in `prow/Makefile`. Please note that I cannot test this, I'm testing locally. I don't know the standars/workflow for creating new commands in `test-infra` and I found no tutorial about this.

alexandercampbell commented on label_sync/Makefile:
```diff
@@ -0,0 +1,59 @@
+test:
+	go test -race -cover $$(go list ./... | grep -v "\/vendor\/")
+
+.PHONY: build test get-cluster-credentials
```
where's get-build-cluster-credentials in this list?

lukaszgryglicki replied:
I suppose it must be called manually. Really I was following `prow/Makefile`. ANd to be honest I need help with deploying this.
Makefile was never tested with actual commands that require `gcloud`, `kubectl` etc.

alexandercampbell commented on label_sync/Makefile:
```diff
@@ -0,0 +1,59 @@
+label_sync-cronjob: get-cluster-credentials
+	kubectl apply -f cluster/label_sync_cron_job.yaml
+
+.PHONY: label_sync-image label_sync-deployment label_sync-cronjob
```
Where's compile_static in this list?

And for what it's worth, convention in Makefiles is to use dashes between names. For example, `compile-static` instead of `compile_static`.

lukaszgryglicki replied:
`compile-static` is defined, and `label_sync-image` depends on it, and `label_sync-image` is part of `all`/`.PHONY`

alexandercampbell commented on label_sync/cluster/label_sync_cron_job.yaml:
```diff
@@ -0,0 +1,29 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+---
```
Is this separator required? I'm just curious.

lukaszgryglicki replied:
It is not required, but YAML Lint wanted it (it was displaying warning)

alexandercampbell commented on label_sync/cluster/label_sync_deployment.yaml:
```diff
@@ -0,0 +1,42 @@
+          image: gcr.io/k8s-prow/label_sync:0.100
+          imagePullPolicy: Always
+          # args:
+          # - -debug
```
Did you mean to include this commented-out config?

lukaszgryglicki replied:
Comment removed

alexandercampbell commented on label_sync/main.go:
```diff
@@ -0,0 +1,439 @@
+	repoLabelsData     = flag.String("repo-labels", "", "Repository Labels file saved as YAML to save GitHub API points")
+	dumpReposData      = flag.String("dump-repos", "", "Save repositories found as YAML (so this file can be used again with -repos without using Gihub API points in -local mode)")
+	dumpRepoLabelsData = flag.String("dump-repo-labels", "", "Save repositories labels found as YAML (so this file can be used again with -repo-labels without using Gihub API points in -local mode)")
+	debug              = flag.Bool("debug", false, "Turn on debug to be more verbose")
```
Generally speaking we've used [Cobra](https://github.com/spf13/cobra) in Kubernetes for command-line applications. I don't necessarily think it's better than the native flag system but it might be worth being consistent. If one of the other reviewers wants to chime in I'd appreciate it.

lukaszgryglicki replied:
Left untouched.

alexandercampbell commented on prow/github/client.go:
```diff
@@ -440,6 +440,72 @@ func (c *Client) CreateStatus(org, repo, ref string, s Status) error {
 }
 
+func (c *Client) GetRepos(org string, isUser bool) ([]Repo, error) {
+	c.log("GetRepos", org, isUser)
```
Thanks Go. lol

alexandercampbell commented on prow/github/client.go:
```diff
@@ -440,6 +440,72 @@ func (c *Client) CreateStatus(org, repo, ref string, s Status) error {
+	if c.fake {
+		repos = append(repos, Repo{Owner: User{Name: "kubernetes"}, Name: "kubernetes", FullName: "kubernetes/kubernetes", HTMLURL: "https://github.com/kubernetes/kubernetes"})
+		repos = append(repos, Repo{Owner: User{Name: "kubernetes"}, Name: "minikube", FullName: "kubernetes/minikube", HTMLURL: "https://github.com/kubernetes/minikube"})
+		repos = append(repos, Repo{Owner: User{Name: "kubernetes"}, Name: "test-infra", FullName: "kubernetes/testinfra", HTMLURL: "https://github.com/kubernetes/test-infra"})
```
There should be a better way to structure this besides adding test-related code in the Client methods themselves. Why don't you have a second client implementation that returns these fake repos?

lukaszgryglicki replied:
Not needed, removed.

spxtr commented on label_sync/main.go:
```diff
@@ -0,0 +1,439 @@
+	repoLabelsData     = flag.String("repo-labels", "", "Repository Labels file saved as YAML to save GitHub API points")
+	dumpReposData      = flag.String("dump-repos", "", "Save repositories found as YAML (so this file can be used again with -repos without using Gihub API points in -local mode)")
+	dumpRepoLabelsData = flag.String("dump-repo-labels", "", "Save repositories labels found as YAML (so this file can be used again with -repo-labels without using Gihub API points in -local mode)")
+	debug              = flag.Bool("debug", false, "Turn on debug to be more verbose")
```
For most of our test infrastructure we're using go's built-in flags. Unless the application is needing cobra's features, I think it's best to stick with the built-in.

kargakis replied:
> For most of our test infrastructure we're using go's built-in flags. Unless the application is needing cobra's features, I think it's best to stick with the built-in.

Agreed.

grodrigues3 commented:
The comment's changes are small (not super important nits), but being able to roll out to a few repos initially is an important configuration option.

grodrigues3 commented:
@lukaszgryglicki ping?

LGTM except an option that let's us roll out to a subset of repositories

lukaszgryglicki commented:
This can be done by specifying custom "repos.yaml" file.
Let me know which repos should be defined in this subset and/or eventually which label.
Once I know that I'll create a file with such repos listed and add it as a parameter to `label_sync` call.

lukaszgryglicki commented on label_sync/main.go:
```diff
@@ -0,0 +1,439 @@
+}
+
+var (
+	labelsPath         = flag.String("labels-path", "/etc/config/labels.yaml", "Path to labels.yaml.")
```
There is such a flag already `reposData`.
You can specify repository list to run there.

lukaszgryglicki commented on label_sync/main.go:
```diff
@@ -0,0 +1,439 @@
+	return nil
+}
+
+// UnmarshalReposData returns GitHub data recorded to YAML file via -dump-repos
```
Yes and You can also use this file to override list of repositories to run this program on, comment updated

grodrigues3 commented:
@spiffxp correct me if this sounds like a mistake.  Also, finalize check that the labels.yaml file is update to date with the subset we want copied

Repos to start with:
- test-infra
- community

grodrigues3 commented:
After those changes are made.  LGTM

lukaszgryglicki commented:
Added `repos.yaml` file that defines restricted run (only on `test-infra` and `community`).
Added copying this file to the container in `Dockerfile`
Added command argument `--repos=/etc/config/repos.yaml` to the `cluster/label_sync_deployment.yaml` deployment.
Also fixed creating GitHub client (`NewClient`, `NewDryRunClient`) in `main.go` - seems like parameter (API endpoint) was added after oauth-secret.

lukaszgryglicki commented on label_sync/repos.yaml:
```diff
@@ -0,0 +1,6 @@
+- full_name: kubernetes/test-infra
+  name: test-infra
+- full_name: kubernetes/community
+  name: community
```
This file (when used as a parameter --repos=repos.yaml) limits list of repositories to run `sync_labels` on.

lukaszgryglicki commented:
This is the result of `dry-run` with current labels defined in `labels.yaml` and repositories limited by `repos.yaml` file:
```
make compile-static
CGO_ENABLED=0 go build -o label_sync k8s.io/test-infra/label_sync
./label_sync --labels-path=labels.yaml --repos=repos.yaml --dry-run=true --debug=true > out.txt
```
[out.txt](https://github.com/kubernetes/test-infra/files/1259220/out.txt)

lukaszgryglicki commented on label_sync/cluster/label_sync_deployment.yaml:
```diff
@@ -0,0 +1,42 @@
+        - name: label_sync
+          image: gcr.io/k8s-prow/label_sync:0.100
+          args:
+          - --repos=/etc/config/repos.yaml
```
Pass custom `repos.yaml` to limit the list of repositories to run `label_sync` on.

lukaszgryglicki commented on label_sync/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+ADD label_sync /label_sync
+RUN mkdir /etc/config
+ADD labels.yaml /etc/config/labels.yaml
+ADD repos.yaml /etc/config/repos.yaml
```
Copy `repos.yaml` to the container.

grodrigues3 commented on label_sync/labels.yaml:
```diff
@@ -0,0 +1,355 @@
+  - color: eb6420
+    name: priority/important-soon
+  - color: ff0000
+    name: priority/P0
```
based on out.txt I think we want to get rid of priority/p0

grodrigues3 replied:
we want to get rid of all the priority/p# labels

lukaszgryglicki commented:
@grodrigues3 all the priority/p# labels commented out.

grodrigues3 commented:
Sorry, one last change that just came up yesterday.  There's a **sig/gcp** being created.

Can we add that label (same color as the others).  After that,

/lgtm

lukaszgryglicki commented:
@grodrigues3 `sig/gcp` added with the same color as other sig labels `d2b48c`.
Run `verify-bazel.sh`, fixed label_sync/BUILD - that should fix CI errors.

stevekuznetsov commented:
@lukaszgryglicki could we please add the following items to `labels.yaml`:

```yaml
- name: do-not-merge/work-in-progress
  color: e11d21
- name: do-not-merge/hold
  color: e11d21
- name: do-not-merge/cherry-pick-not-approved
  color: e11d21
- name: do-not-merge/release-note-label-needed
  color: e11d21
- name: do-not-merge/blocked-paths
  color: e11d21
```

lukaszgryglicki commented:
@stevekuznetsov done.

lukaszgryglicki commented:
/retest

stevekuznetsov commented:
@lukaszgryglicki thanks!

/retest

lukaszgryglicki commented:
Tests passed, merge?

kargakis commented on prow/github/client.go:
```diff
@@ -475,11 +538,13 @@ func (c *Client) GetCombinedStatus(org, repo, ref string) (*CombinedStatus, erro
+	var labels []Label
 	if c.fake {
-		return nil, nil
+		labels = append(labels, Label{Name: "lgtm", Color: "green"})
```
@spxtr having test code inside prod is very very ugly. Why don't we follow the way fake clients are built in k8s, that is, by using a client interface?

lukaszgryglicki replied:
LOL I thought that I've removed this....

lukaszgryglicki replied:
I've tested locally with those test data removed:
```
iMac-Adm:label_sync lukasz$ git diff
diff --git a/prow/github/client.go b/prow/github/client.go
index 31de8e9a..690bf254 100644
--- a/prow/github/client.go
+++ b/prow/github/client.go
@@ -540,8 +540,6 @@ func (c *Client) GetCombinedStatus(org, repo, ref string) (*CombinedStatus, erro
 func (c *Client) getLabels(path string) ([]Label, error) {
        var labels []Label
        if c.fake {
-               labels = append(labels, Label{Name: "lgtm", Color: "green"})
-               labels = append(labels, Label{Name: "priority/P0", Color: "red"})
                return labels, nil
        }
        nextURL := strings.Join([]string{c.base, path}, "")
iMac-Adm:label_sync lukasz$ 
```

This is not needed for tests and can be removed any time:
```
iMac-Adm:label_sync lukasz$ make test
go test -race -cover $(go list ./... | grep -v "\/vendor\/")
ok  	k8s.io/test-infra/label_sync	1.074s	coverage: 32.7% of statements
iMac-Adm:label_sync lukasz$ 
```

kargakis replied:
Mind opening a follow-up pr to remove them?

lukaszgryglicki replied:
I need to finish for today.
I'll open tomorrow unless anybody else does it until then.
It is just a super trivial PR to open 2 lines of code in a single file.

lukaszgryglicki replied:
OK I'm on it now

stevekuznetsov commented:
Once we run the sync and get the labels created we should be in good shape to merge #4212

lukaszgryglicki commented:
Thanks!

lukaszgryglicki commented:
@stevekuznetsov test code removed from prow/github/client.go, see this comment: https://github.com/kubernetes/test-infra/pull/2706#discussion_r136496229
